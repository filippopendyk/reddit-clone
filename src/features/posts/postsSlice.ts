import { createAsyncThunk, createSlice, PayloadAction } from "@reduxjs/toolkit";

type RedditApiResponse = {
    data: {
        children: {
            data: {
                ups: number;
                created: number;
            }
        }[];
    }
};

export type RedditPost = {
    ups: number;
    created: number;
}

type FetchPostsPayload = {
    posts: RedditPost[];
    topic: string;
}

export const fetchPosts = createAsyncThunk<FetchPostsPayload, string>(
    'posts/fetchPosts', async(topic: string, thunkApi) => {
        try{
            //Fetch posts from specified category
            const response = await fetch(`https://www.reddit.com/r/${topic}.json`);
            const data: RedditApiResponse = await response.json();

            //Parse the response data
            const posts = data.data.children.map((child) => child.data);

            return { posts, topic };
        } catch (err) {
            throw new Error('failed to fetch posts');
        }
    }
)

type PostsState = {
    isLoading: boolean;
    error: string | null;
    data: RedditPost[];
}

const initialState: PostsState = {
    isLoading: false,
    error: null,
    data: [],
}

const postsSlice = createSlice({
    name: 'posts',
    initialState,
    reducers: {},
    extraReducers: (builder) => {
        builder.addCase(fetchPosts.pending, (state) => {
            state.isLoading = true;
            state.error = null;
        });
        builder.addCase(fetchPosts.fulfilled, (state, action: PayloadAction<FetchPostsPayload>) => {
            state.isLoading = false;
            state.data = action.payload.posts;
        });
        builder.addCase(fetchPosts.rejected, (state, action) => {
            state.isLoading = false;
            state.error = action.error.message || 'Something went wrong';
        });
    }
}
);

export const { reducer: postsSliceReducer } = postsSlice;