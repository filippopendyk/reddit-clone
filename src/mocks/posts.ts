const posts = [
    {
        ups: 123,
        created: 123456789,
        title: 'example',
        id: 123,
        author: 'filip popendyk',
        subreddit_name_prefixed: 'r/example',
        thumbnail: 'https://i.redd.it/ccskoy6lvx881.jpg',
        upvote_ratio: 0.97,
        selftext: 'siemano'
    },
    {
        ups: 2137,
        created: 21372134337,
        title: '2137',
        id: 2137,
        author: 'wykop.pl',
        subreddit_name_prefixed: 'r/2137',
        thumbnail: '',
        upvote_ratio: 0.96,
        selftext: ''
    }
]

export default posts;