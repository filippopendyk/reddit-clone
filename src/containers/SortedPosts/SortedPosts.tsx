import { RedditPost } from "../../features/posts/postsSlice";
import Post from "../../components/Post/Post";

type Props = {
    filteredPosts: RedditPost[]
}

const SortedPosts: React.FC<Props> = ({filteredPosts}) => {
    console.log(filteredPosts);
    
    return (
        <ul>
            {
                filteredPosts.map((post) => {
                    return <Post post={post} key={post.id} id={post.id} author={post.author} thumbnail={post.thumbnail} subreddit_name_prefixed={post.subreddit_name_prefixed}/>
                })
            }
        </ul>
    )
}

export default SortedPosts;