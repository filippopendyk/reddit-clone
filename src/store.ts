import { configureStore } from "@reduxjs/toolkit";
import filterSlice from "./features/filter/filterSlice";
import { postsSliceReducer } from "./features/posts/postsSlice";
import { fetchMostPopularCategories, topCategoriesReducer } from "./features/topCategories/topCategoriesSlice";

const store = configureStore({
    reducer: {
        topCategories: topCategoriesReducer,
        posts: postsSliceReducer,
        filter: filterSlice,
    },
});

store.dispatch(fetchMostPopularCategories());

export default store;

// Define types for usedispatch and useselector

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch;