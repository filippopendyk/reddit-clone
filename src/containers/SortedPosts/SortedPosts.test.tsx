import SortedPosts from "./SortedPosts";
import renderer from 'react-test-renderer';
import { Provider } from "react-redux";
import store from "../../store";
import posts from "../../mocks/posts";
import { render, screen } from "@testing-library/react";

describe('SortedPosts', () => {
    it('Renders without crashing', () => {
        const tree = renderer
        .create(
            <Provider store={store}>
                <SortedPosts filteredPosts={posts} />
            </Provider>
        )
        .toJSON();
        expect(tree).toMatchSnapshot();
    })

    it('Renders list of posts', () => {
        render(
            <Provider store={store}>
                <SortedPosts filteredPosts={posts}/>
            </Provider>
        )

        const listOfPosts = screen.getByRole('list');
        expect(listOfPosts).toBeInTheDocument();
    })
})