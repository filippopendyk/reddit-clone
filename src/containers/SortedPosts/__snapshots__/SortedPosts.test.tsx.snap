// Vitest Snapshot v1

exports[`SortedPosts > Renders without crashing 1`] = `
<ul>
  <li>
    <div
      className="stats-container"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 320 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M177 159.7l136 136c9.4 9.4 9.4 24.6 0 33.9l-22.6 22.6c-9.4 9.4-24.6 9.4-33.9 0L160 255.9l-96.4 96.4c-9.4 9.4-24.6 9.4-33.9 0L7 329.7c-9.4-9.4-9.4-24.6 0-33.9l136-136c9.4-9.5 24.6-9.5 34-.1z"
        />
      </svg>
      <p>
        123
      </p>
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 320 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
        />
      </svg>
    </div>
    <div
      className="content-container"
    >
      <img
        alt="example"
        src="https://i.redd.it/ccskoy6lvx881.jpg"
      />
      <article
        className="text-container"
      >
        <p
          className="subreddit-name"
        >
          r/example
        </p>
        <p
          className="author-time"
        >
          Posted by 
          filip popendyk
           
          49 years ago
           
        </p>
        <p
          className="title"
        >
          example
        </p>
        <p
          className="selftext"
          data-testid="selftext"
        >
          siemano
        </p>
      </article>
    </div>
  </li>
  <li>
    <div
      className="stats-container"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 320 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M177 159.7l136 136c9.4 9.4 9.4 24.6 0 33.9l-22.6 22.6c-9.4 9.4-24.6 9.4-33.9 0L160 255.9l-96.4 96.4c-9.4 9.4-24.6 9.4-33.9 0L7 329.7c-9.4-9.4-9.4-24.6 0-33.9l136-136c9.4-9.5 24.6-9.5 34-.1z"
        />
      </svg>
      <p>
        2k
      </p>
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 320 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
        />
      </svg>
    </div>
    <div
      className="content-container"
    >
      <article
        className="text-container"
      >
        <p
          className="subreddit-name"
        >
          r/2137
        </p>
        <p
          className="author-time"
        >
          Posted by 
          wykop.pl
           
          -19693860359 seconds ago
           
        </p>
        <p
          className="title"
        >
          2137
        </p>
      </article>
    </div>
  </li>
</ul>
`;
